### PR TITLE
Fk support

### PR DESCRIPTION
**Pull request by srinidhis94**

### CODE REVIEW

File: drivers/sqlcommons.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+			}
+		}
+		tableDescriptorMap[table] = fields
+	}
+	return tableDescriptorMap, newlyReferencedTables, nil
+}
+
+func getInsertionOrder(tablesToFieldsMap map[string][]FieldDescriptor) ([]string, error) {
+	var tablesVisitOrder []string
+	tablesVisited := make(map[string]bool)
```
Comment (borosr) [inline drivers/sqlcommons.go:40]: This can be more efficient if you're using `map[string]struct{}` instead of `map[string]bool`.

File: drivers/sqlcommons.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+				break
+			}
+			if canInsert {
+				newInsertCount++
+				tablesVisited[table] = true
+				tablesVisitOrder = append(tablesVisitOrder, table)
+			}
+		}
+		if newInsertCount == 0 {
+			break
```
Comment (borosr) [inline drivers/sqlcommons.go:66]: Can't you return here with `errors.New("error generating insertion order. Maybe necessary dependencies are not met")`?

File: drivers/sqlcommons.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	return tablesVisitOrder, nil
+}
+
+func testTable(db *sql.DB, testCase, table string, d Driver) error {
+	test, err := d.GetTestCase(testCase)
+	if err != nil {
+		return err
+	}
+	if test.TableCreationOrder == nil {
+		if query, ok := test.TableToCreateQueryMap[""]; ok {
```
Comment (borosr) [inline drivers/sqlcommons.go:81]: This empty string key should be a package level, exported constant and name it please `DefaultTableCreateQueryKey` if I understand it right, and use it in the test instance declaration as well.

File: pkg/action/action.go [hunk 4]
```diff
@@ -17,8 +17,58 @@ import (
...
+				return err
+			}
+		}
+	}
+	return nil
+}
+
 // Insert is inserting a random generated data into the chosen table
-func Insert(db *sql.DB, fields []drivers.FieldDescriptor, driver drivers.Driver, table string) error {
+func Insert(args ...interface{}) error {
```
Comment (borosr) [inline pkg/action/action.go:66]: Why did you change this to be that generic?

File: pkg/fuzzer/runner.go [hunk 5]
```diff
@@ -31,6 +39,57 @@ func Run(fields []drivers.FieldDescriptor, f flags.Flags) error {
...
+// Run the commands in a worker pool
+func Run(fields []drivers.FieldDescriptor, f flags.Flags) error {
+	driver, db := getDriverAndDB(f)
+	defer func() {
+		if err := db.Close(); err != nil {
+			log.Print(err)
+		}
+	}()
+	return runHelper(f, action.Insert, []interface{}{db, fields, driver, f.Table})
+	//numJobs := f.Num
```
Comment (borosr) [inline pkg/fuzzer/runner.go:51]: Please remove the comments

**Review by borosr — CHANGES_REQUESTED** [review]

File: pkg/action/action.go (hunk 4 above)
Comment (srinidhis94) [reply pkg/action/action.go:66]: This change was to use same worker pool logic for both insert and multi insert .  But  I also discussed this with @PumpkinSeed and one more solution i can think of is  something like this.
type insertActionStruct struct{
    insertArgs *Insert
    multiInsertArgs *MultiInsert
}

and 
func (ias *insertActionStruct) insert(){
}

and
func runHelper(f flags.Flags, ias insertActionStruct) error {
...
...
}
If there's any other alternative, let me know.

File: pkg/connector/connector.go [hunk 6]
```diff
@@ -15,7 +16,9 @@ func Connection(d drivers.Driver) *sql.DB {
 		log.Fatal(err)
 		return nil
 	}
-
+	db.SetConnMaxLifetime(100 * time.Second)
```
Comment (PumpkinSeed) [inline pkg/connector/connector.go:19]: These should be coming from flags.
```
type Flags struct {
	Driver drivers.Flags

	Num     int
	Workers int
	Table   string
	ConnMaxLifetime int
	MaxIdleConns int
	MaxOpenConns int

	Parsed  bool
}
```

File: pkg/fuzzer/runner.go [hunk 7]
```diff
@@ -11,15 +12,22 @@ import (
 	_ "github.com/lib/pq"
 )
 
-// Run the commands in a worker pool
-func Run(fields []drivers.FieldDescriptor, f flags.Flags) error {
+func getDriverAndDB(f flags.Flags) (drivers.Driver, *sql.DB) {
+	driver := drivers.New(f.Driver)
+	db := connector.Connection(driver)
+	db.SetMaxOpenConns(f.Workers)
```
Comment (PumpkinSeed) [inline pkg/fuzzer/runner.go:18]: If it's set in the connector, don't set it here. Pick one place to set it. I guess this is easier here. But create new flag options for these.

File: pkg/action/action.go (hunk 4 above)
Comment (PumpkinSeed) [reply pkg/action/action.go:66]: What you can do here is creating a struct like that:

```
type InsertEnum struct {
  Single *SingleInsertParams
  Multi   *MultiInsertParams
}
```

If you pass nil MultiInsertParams into the MultiInsert it will throw en error. In this case the programmer will know what parameters should have been passed to the underlying function and we have a kind of type safety.

**Review by PumpkinSeed — CHANGES_REQUESTED** [review]

**Comment by PumpkinSeed** [comment]
Also I faced with an issue. In the pkg/action/action.go can you change the 131. lines to this: `gofakeit.Number(1980, 2028),` please?

**Review by PumpkinSeed — APPROVED** [review]
